state-chain/pallets/cf-trading-strategy/src/lib.rs

```diff
@@ -0,0 +1,351 @@
+			base_asset: Asset,
+			strategy: TradingStrategy,
+		},
+		FundsAddedToStrategy {
+			strategy_id: T::AccountId,
+			base_asset: Asset,
+			base_asset_amount: AssetAmount,
+			quote_asset_amount: AssetAmount,
+		},
+		StrategyClosed {
+			strategy_id: T::AccountId,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		StrategyNotFound,
+		AmountBelowDeploymentThreshold,
+		InvalidOwner,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(1)]
+		#[pallet::weight(Weight::zero())] // TODO: benchmark
+		pub fn deploy_trading_strategy(
+			origin: OriginFor<T>,
+			base_asset_amount: AssetAmount,
+			quote_asset_amount: AssetAmount,
+			base_asset: Asset,
+			strategy: TradingStrategy,
+		) -> DispatchResult {
+			let lp = &T::AccountRoleRegistry::ensure_liquidity_provider(origin)?;
+
+			let quote_asset = STABLE_ASSET;
+
+			let strategy_id = {
+				// Check that strategy is created with sufficient funds:
+				{
+					const ONE: FixedU64 = FixedU64::from_u32(1);
+					let parameters = TradingStrategyParametersStorage::<T>::get();
+					let fraction_of_required = |asset, provided| {
+						let required = parameters.get_strategy_deployment_threshold(&asset);
+
+						if required == 0 {
+							ONE
+						} else {
+							FixedU64::from_rational(provided, required)
+						}
+					};
+
+					ensure!(
+						fraction_of_required(base_asset, base_asset_amount) +
+							fraction_of_required(quote_asset, quote_asset_amount) >=
+							ONE,
+						Error::<T>::AmountBelowDeploymentThreshold
+					);
+				}
+
+				let nonce = T::NonceProvider::get_nonce(lp);
+
+				let strategy_id = derive_strategy_id::<T>(lp, nonce);
+
+				Self::deposit_event(Event::<T>::StrategyDeployed {
+					account_id: lp.clone(),
+					strategy_id: strategy_id.clone(),
+					base_asset,
+					strategy: strategy.clone(),
+				});
+
+				Strategies::<T>::insert(
+					strategy_id.clone(),
+					TradingStrategyEntry { base_asset, owner: lp.clone(), strategy },
+				);
+
+				strategy_id
+			};
+
+			Self::add_funds_to_existing_strategy(
+				lp,
+				&strategy_id,
+				base_asset,
+				base_asset_amount,
+				quote_asset_amount,
+			)
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(Weight::zero())] // TODO: benchmark
+		pub fn close_strategy(origin: OriginFor<T>, strategy_id: T::AccountId) -> DispatchResult {
+			let lp = &T::AccountRoleRegistry::ensure_liquidity_provider(origin)?;
+
+			let TradingStrategyEntry { base_asset, owner, strategy } =
+				Strategies::<T>::take(&strategy_id).ok_or(Error::<T>::StrategyNotFound)?;
+
+			ensure!(lp == &owner, Error::<T>::InvalidOwner);
+
+			// TODO: instead of reading ticks from the strategy, we could extend PoolApi with
+			// a method to close all (limit) orders (which might be necessary for more complex
+			// strategies).
+			let TradingStrategy::SellAndBuyAtTicks { buy_tick, sell_tick } = strategy;
+
+			let cancel_limit_orders = |side, tick| {
+				T::PoolApi::update_limit_order(
```

Comment:
Seems like the PoolApi could do with a close_limit_order method...

Reply:
Created a `close_limit_order`